File: crack-the-coding-interview/linked_lists/remove_dups.c
/*
	Remove Dups! Write code to remove duplicates from an unsorted linked list.
	FOLLOW UP
	How would you solve this problem if a temporary bu er is not allowed?
*/


#include<stdio.h>
#include<stdlib.h>


struct Node{
	int data;
	struct Node* next;
};



struct Node* NewNode(int data){
	struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
	new_node->data = data;
	new_node->next = NULL;
	return new_node;
}


void printLL(struct Node* head){
	struct Node* p = head;
	while(p->next != NULL){
		printf("%d ", p->data);
		p = p->next;
	}
}



struct Node* removeDups(struct Node* head){
	// brute force algorithm.compare each node with rest of it
	struct Node* current = head;
	struct Node* next = head;

	if current == NULL || current->next == NULL{
		// empty or single linkedlist node
		return head;
	}
	next = current->next;
	while(current->data != next->data){
		next = current->next;
		if(next == NULL){
			// reached end of the LL
			next = current->next;
			current = current->next;
		}
	}


}


int main(){
	struct Node* head = NewNode(1);
	head->next = NewNode(2);
	head->next->next = NewNode(2);
	head->next->next->next = NewNode(3);
	head->next->next->next->next = NewNode(3);

	printLL(head);
}

File: crack-the-coding-interview/linked_lists/remove_dups.py
"""
    Remove Dups! Write code to remove duplicates from an unsorted linked list.
    FOLLOW UP
    How would you solve this problem if a temporary bu er is not allowed?
"""


def remove_dups_with_set(head):
    """
        Complexity: O(n)
        Space Complexity: O(n) for set
    """
    global_set = set()

    current = head
    prev = None
    while current.next is not None:
        if current.data in global_set:
            # already visited node.remove this
            prev.next = current.next
        else:
            global_set.add(current.data)
        prev = current
        current = current.next


def remove_dups(head):
    """
        Complexity is O(n2)
    """
    current = head
    runner = head

    while current.next is not None:
        runner = current.next
        while runner.next is not None:
            if current.data == runner.data:
                current.next = runner.next
            runner = runner.next
        current = current.next


class LinkedList(object):
    def __init__(self, data):
        self.data = data
        self.next = None


def print_ll(head):

    while head is not None:
        print head.data,
        head = head.next

if __name__ == "__main__":
    head = LinkedList(1)
    head.next = LinkedList(2)
    head.next.next = LinkedList(2)
    head.next.next.next = LinkedList(3)
    head.next.next.next.next = LinkedList(3)
    remove_dups_with_set(head)
    print_ll(head)
    print "\n"

    head = LinkedList(1)
    head.next = LinkedList(2)
    head.next.next = LinkedList(2)
    head.next.next.next = LinkedList(3)
    head.next.next.next.next = LinkedList(3)
    remove_dups(head)
    print_ll(head)
